### ISSUE TITLE
Improving memory usage for longer regions

### ISSUE THREAD

**Issue by chapmanb** [AstraZeneca-NGS/VarDictJava#12]
We've been testing VarDictJava on whole genome analyses chunked into analyzable sections. We typically break up the genome at natural breakpoints without coverage to avoid accidentally splitting an analysis across a larger indel or multi-nucleotide change.

This works for 99% of cases but we end up with some regions where we have slightly larger regions (2-5Mb) that cause out of memory errors with VarDictJava. VarDict also struggles with these larger chunks. Thanks are due to @lpantano for diagnosing the underlying issue.

Would it be possible to adjust memory usage so we can work with these larger regions without needing to drastically increase memory? @zhongwulai and @jabbarish might have ideas about places where we can reset some shared state and keep memory usage more stable. Our target has been to do regions up to 5Mb in 3Gb of memory.

Here is a reproducible test case with a small shell script that shows the issue:

https://s3.amazonaws.com/chapmanb/az/vardict_memory.tar.gz

The original region that causes memory issues is:
```
22	16050035	19050035
```
Chunking it into 1Mb blocks lets it finish without any issues on the same memory:
```
22	16050035	17050035
22	17050035	18050035
22	18050035	19050035
```
Here is the memory issue we see -- just a straight up out of memory error:
```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.regex.Pattern.compile(Pattern.java:1655)
        at java.util.regex.Pattern.<init>(Pattern.java:1337)
        at java.util.regex.Pattern.compile(Pattern.java:1022)
        at com.astrazeneca.vardict.VarDict.findMSI(VarDict.java:3126)
        at com.astrazeneca.vardict.VarDict.toVars(VarDict.java:2899)
        at com.astrazeneca.vardict.VarDict.somaticNotParallel(VarDict.java:355)
        at com.astrazeneca.vardict.VarDict.nonAmpVardict(VarDict.java:251)
        at com.astrazeneca.vardict.VarDict.start(VarDict.java:67)
        at com.astrazeneca.vardict.Main.run(Main.java:137)
        at com.astrazeneca.vardict.Main.main(Main.java:24)
```
Thanks for any thoughts or suggestions.

**Comment by zhongwulai** [comment]
Hi Brad,
Thanks for the feedback.  It would require some substantial redesign of the algorithm, which I haven't had a clear idea.  The current algorithm uses memory proportionally to the BED segment sizes.  As the only potential problem chopping segments is losing the indels spanning the boundary, I think it's probably better to create overlapping (>150bp) segments (<1Mb) for now.  That's typically what I did for WGS.  The downstream analysis (converting to VCF) will remove any redundant calls in those 150bp.  For TCGA analysis, I just used gene boundaries so that any large indels within the gene won't be missed.
My current working version will abandon the indel size limits all together and should work just fine for indels even far outside the segments.
Thanks,Zhongwu

 


     On Friday, May 15, 2015 6:15 AM, Brad Chapman <notifications@github.com> wrote:
   
 

 We've been testing VarDictJava on whole genome analyses chunked into analyzable sections. We typically break up the genome at natural breakpoints without coverage to avoid accidentally splitting an analysis across a larger indel or multi-nucleotide change.This works for 99% of cases but we end up with some regions where we have slightly larger regions (2-5Mb) that cause out of memory errors with VarDictJava. VarDict also struggles with these larger chunks. Thanks are due to @lpantano for diagnosing the underlying issue.Would it be possible to adjust memory usage so we can work with these larger regions without needing to drastically increase memory? @zhongwulai and @jabbarish might have ideas about places where we can reset some shared state and keep memory usage more stable. Our target has been to do regions up to 5Mb in 3Gb of memory.Here is a reproducible test case with a small shell script that shows the issue:https://s3.amazonaws.com/chapmanb/az/vardict_memory.tar.gzThe original region that causes memory issues is:22  16050035    19050035
Chunking it into 1Mb blocks lets it finish without any issues on the same memory:22  16050035    17050035
22  17050035    18050035
22  18050035    19050035
Here is the memory issue we see -- just a straight up out of memory error:Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.regex.Pattern.compile(Pattern.java:1655)
        at java.util.regex.Pattern.<init>(Pattern.java:1337)
        at java.util.regex.Pattern.compile(Pattern.java:1022)
        at com.astrazeneca.vardict.VarDict.findMSI(VarDict.java:3126)
        at com.astrazeneca.vardict.VarDict.toVars(VarDict.java:2899)
        at com.astrazeneca.vardict.VarDict.somaticNotParallel(VarDict.java:355)
        at com.astrazeneca.vardict.VarDict.nonAmpVardict(VarDict.java:251)
        at com.astrazeneca.vardict.VarDict.start(VarDict.java:67)
        at com.astrazeneca.vardict.Main.run(Main.java:137)
        at com.astrazeneca.vardict.Main.main(Main.java:24)
Thanks for any thoughts or suggestions.—
